packages/offix-client/src/conflicts/ConflictLink.ts

```diff
@@ -0,0 +1,178 @@
+import { onError, ErrorResponse } from "apollo-link-error";
+import { GraphQLError, ExecutionResult, subscribe } from "graphql";
+import { ApolloLink, Operation, NextLink, Observable, FetchResult } from "apollo-link";
+import { ConflictResolutionData } from "./strategies/ConflictResolutionData";
+import { isMutation } from "../utils/helpers";
+import { ObjectState, ConflictListener } from ".";
+import { ConflictResolutionStrategy } from "./strategies/ConflictResolutionStrategy";
+import { clientWins } from "./strategies/strategies";
+import { ConflictHandler } from "./handler/ConflictHandler";
+
+/**
+ * Local conflict thrown when data outdates even before sending it to the server.
+ * Can be used to correct any data in flight or shown user another UI to visualize new state
+ */
+export class LocalConflictError extends Error {
+  constructor(private base: any, private variables: any) {
+    super();
+  }
+}
+
+/**
+ * Represents conflict information that was returned from server
+ */
+export interface ConflictInfo {
+  serverState: ConflictResolutionData;
+  clientState: ConflictResolutionData;
+  // Expected return type of the mutation
+  returnType: string;
+}
+
+/**
+ * Configuration for conflict resolution
+ */
+export interface ConflictConfig {
+  /**
+   * Interface that defines how object state is progressed
+   * This interface needs to match state provider supplied on server.
+   */
+  conflictProvider: ObjectState;
+
+  /**
+   * Interface that can be implemented to receive information about the data conflict
+   *
+   * @deprecated see OfflineClient.registerOfflineEventListener
+   */
+  conflictListener?: ConflictListener;
+
+  /**
+   * The conflict resolution strategy your client should use. By default it takes client version.
+   */
+  conflictStrategy?: ConflictResolutionStrategy;
+}
+
+/**
+ * Conflict handling link implementation that provides ability to determine
+ */
+export class ConflictLink extends ApolloLink {
+  private stater: ObjectState;
+  private strategy: ConflictResolutionStrategy | undefined;
+  private listener: ConflictListener | undefined;
+
+  constructor(private config: ConflictConfig) {
+    super();
+    this.stater = this.config.conflictProvider;
+    this.strategy = this.config.conflictStrategy;
+    this.listener = this.config.conflictListener;
+  }
+
+  public request(
+    operation: Operation,
+    forward: NextLink
+  ): Observable<FetchResult> | null {
+    if (isMutation(operation)) {
+      if (this.stater.currentState(operation.variables) !== undefined) {
+        return new Observable(observer => {
+          let retriedResult: any;
+          let sub: any;
+          let retriedSub: any;
+          sub = forward(operation).subscribe({
```

Comment:
This code is ok, but it is quite hard to read and unit test.
If we extract observable logic into a separate class (retried observable or something). 
It is fine to leave it but mixing observables and business logic is why Apollo Client source code is really hard to read and understand.

Reply:
I actually figured out how to use onError for our specific case instead of just dropping the code in here like this. Also makes it way easier to test. Pushing soon.